src/stan/model/model_base_crtp.hpp

```diff
@@ -0,0 +1,216 @@
+#include <vector>
+
+namespace stan {
+namespace model {
+
+/**
+ * Base class employing the curiously recursive template pattern for
+ * static inheritance to adapt templated `log_prob` and `write_array`
+ * methods to their untemplated virtual counterparts declared in
+ * `model_base`.
+ *
+ * The derived class `M` is required to implement the following two
+ * pairs of template functions,
+ *
+ * ```
+ * template <bool propto, bool jacobian, typename T>
+ * T log_prob(std::vector<T>& params_r,
+ *              std::vector<int>& params_i,
+ *              std::ostream* msgs = 0) const;
+
+ * template <bool propto, bool jacobian, typename T>
+ * T log_prob(Eigen::Matrix<T, -1, 1>& params_r,
+ *            std::ostream* msgs = 0) const;
+ * ```
+ *
+ * and
+ *
+ * ```
+ *  template <typename RNG>
+ *    void write_array(RNG& base_rng,
+ *                     std::vector<double>& params_r,
+ *                     std::vector<int>& params_i,
+ *                     std::vector<double>& vars,
+ *                     bool include_tparams = true,
+ *                     bool include_gqs = true,
+ *                     std::ostream* msgs = 0) const;
+ *
+ *   template <typename RNG>
+ *    void write_array(RNG& base_rng,
+ *                     Eigen::Matrix<double, -1, 1>& params_r,
+ *                     Eigen::Matrix<double, -1, 1>& vars,
+ *                     bool include_tparams = true,
+ *                     bool include_gqs = true,
+ *                     std::ostream* msgs = 0) const
+
+ * ```
+ *
+ * <p>The derived class `M` must be declared following the curiously
+ * recursive template pattern, for example, if `M` is `foo_model`,
+ * then `foo_model` should be declared as
+ *
+ * ```
+ * class foo_model : public stan::model::model_base_crtp<foo_model> { ... };
+ * ```
+ *
+ * The so-called "curious recursion involves the class type that is
+ * being declared (`foo_model`) showing up as a template parameter in
+ * the class it extends (`model_base_crtp<foo_model>`).  In general, the
+ * template parameter `M` for this class is called the derived type,
+ * and must be declared to extend `foo_model<M>`.
+ *
+ * @tparam M type of derived model, which must implemented the
+ * template methods defined in the class documentation
+ */
+template <typename M>
+class model_base_crtp : public stan::model::model_base {
+ public:
+  /**
+   * Construct a model with the specified number of real unconstrained
+   * parameters.
+   *
+   * @param[in] num_params_r number of real unconstrained parameters
+   */
+  explicit model_base_crtp(size_t num_params_r) :
+      model_base(num_params_r) { }
+
+  /**
+   * Destroy this class.  This is required to be virtual to allow
+   * subclass references to clean up superclasses, but is otherwise a
+   * no-op.
+   */
+  virtual ~model_base_crtp() { }
+
+  inline double log_prob(Eigen::VectorXd& theta,
+                         std::ostream* msgs) const override {
+    return static_cast<const M*>(this)
```

Comment:
Would it be possible to declare a reference to the parent class once? So you could have the member

`const M& model_;`

This would be initialized with

`model_(static_cast<const M&>(*this))`

With that you can avoid all those `static_cast` calls and simply do `model_.template ...`.

That's easier to read to me. Does that make sense?

Reply:
Ok... if that's the idiomatic CRTP way.. fine.